File: 2025/10/10_2_leetcode_bc/124.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    let best = -Infinity;

    const getSum = (node) => {
        if (!node) return 0;

        let left = Math.max(0, getSum(node.left));
        let right = Math.max(0, getSum(node.right));
        let curMax = node.val + left + right;
        best = Math.max(best, curMax);

        return node.val + Math.max(left, right);
    }

    getSum(root);

    return best
};

File: 2025/10/10_2_leetcode_bc/124.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        best = float("-inf")

        def getSum(root):
            nonlocal best

            if not root:
                return 0

            left = max(0, getSum(root.left))
            right = max(0, getSum(root.right))
            best = max(best, root.val + left + right)
            return root.val + max(left, right)

        getSum(root)
        return best


